### PR TITLE
Speed up Random Battles species selection

### PR DESCRIPTION
**Pull request by livid-washed**
This refactors species selection in Random Battles to be more efficient. 

Previously, after selecting a base species, it goes through the entire array of species to record the ones with the corresponding base species, then chooses one of them randomly.

Now, `getPokemonPool()` outputs an object `pokemonPool` which records, for each base species, the list of species with that base species. Instead of checking every species, it now just samples from the list corresponding to the base species.

This approach is significantly faster and keeps the appearance rates standardization from https://github.com/smogon/pokemon-showdown/pull/9708.

My local testing shows that this speeds up overall team generation by around 20% for Gen 9 Random Battle and 50% for Gen 7 Random Battle.

### CODE REVIEW

**Review by ACakeWearingAHat — APPROVED** [review]
neat, speed

File: data/random-teams.ts [hunk 1]
```diff
@@ -1731,18 +1730,37 @@ export class RandomTeams {
...
 			}
-			pokemonPool.push(pokemon);
-			baseSpeciesCount[species.baseSpecies] = (baseSpeciesCount[species.baseSpecies] || 0) + 1;
+
+			if (species.baseSpecies in pokemonPool) {
+				pokemonPool[species.baseSpecies].count++;
+				pokemonPool[species.baseSpecies].speciesArray.push(species.id);
+			} else {
+				pokemonPool[species.baseSpecies] = {count: 1, speciesArray: [species.id]};
+			}
```
Comment (urkerab) [inline data/random-teams.ts:1739]: At this point `count` is always the same as `speciesArray.length`...
```suggestion
			if (species.baseSpecies in pokemonPool) {
				pokemonPool[species.baseSpecies].push(species.id);
			} else {
				pokemonPool[species.baseSpecies] = [species.id];
			}
```

File: data/random-teams.ts [hunk 2]
```diff
@@ -1731,18 +1730,37 @@ export class RandomTeams {
...
+			/**
+			 * Weighted random shuffle
+			 * Uses the fact that for two uniform variables x1 and x2, x1^(1/w1) is larger than x2^(1/w2)
+			 * with probability equal to w1/(w1+w2), which is what we want. See e.g. here https://arxiv.org/pdf/1012.0256.pdf,
+			 * original paper is behind a paywall.
+			 */
+			const sortObject = {
+				baseSpecies: baseSpecies,
+				score: Math.pow(this.prng.next(), 1 / pokemonPool[baseSpecies].count),
+			};
```
Comment (urkerab) [inline data/random-teams.ts:1759]: ... where you only use it to calculate the `score` here.
```suggestion
			const sortObject = {
				baseSpecies: baseSpecies,
				score: Math.pow(this.prng.next(), 1 / weight),
			};
```

File: data/random-teams.ts [hunk 3]
```diff
@@ -1716,11 +1716,10 @@ export class RandomTeams {
 		pokemonToExclude: RandomTeamsTypes.RandomSet[] = [],
 		isMonotype = false,
 		pokemonList: string[]
-	) {
+	): [{[k: string]: {count: number, speciesArray: ID[]}}, {baseSpecies: string, score: number}[]] {
 		const exclude = pokemonToExclude.map(p => toID(p.species));
-		const pokemonPool = [];
-		const baseSpeciesPool = [];
-		const baseSpeciesCount: {[k: string]: number} = {};
+		const pokemonPool: {[k: string]: {count: number, speciesArray: ID[]}} = {};
```
Comment (urkerab) [inline data/random-teams.ts:1721]:
```suggestion
		const pokemonPool: {[k: string]: ID[]} = {};
```

File: data/random-teams.ts [hunk 4]
```diff
@@ -1774,17 +1792,14 @@ export class RandomTeams {
...
-			for (const poke of pokemonPool) {
-				const species = this.dex.species.get(poke);
-				if (species.baseSpecies === baseSpecies) currentSpeciesPool.push(species);
-			}
-			let species = this.sample(currentSpeciesPool);
+
+		while (shuffledBaseSpecies.length && pokemon.length < this.maxTeamSize) {
+			// repeated popping from weighted shuffle is equivalent to repeated weighted sampling without replacement
+			const baseSpecies = shuffledBaseSpecies.pop()!.baseSpecies;
+			let species = this.dex.species.get(this.sample(pokemonPool[baseSpecies].speciesArray));
```
Comment (urkerab) [inline data/random-teams.ts:1802]: (Obviously there are some other fixes required; these suggestions are not meant to be exhaustive.)
```suggestion
			let species = this.dex.species.get(this.sample(pokemonPool[baseSpecies]));
```

File: data/random-teams.ts [hunk 5]
```diff
@@ -1731,18 +1730,37 @@ export class RandomTeams {
...
-				// Squawkabilly has 4 formes, but only 2 functionally different formes, so only include it 1x
-				if (baseSpecies === 'Squawkabilly') break;
-			}
+		for (const baseSpecies of Object.keys(pokemonPool)) {
+			// Squawkabilly has 4 formes, but only 2 functionally different formes, so only include it 1x
+			if (baseSpecies === 'Squawkabilly') {
+				pokemonPool[baseSpecies].count = 1;
+			} else {
+				pokemonPool[baseSpecies].count = Math.min(Math.ceil(pokemonPool[baseSpecies].count / 3), 3);
+			}
```
[the remark this answers is not in the document]
Comment (urkerab) [reply data/random-teams.ts]: Oops, thanks for noticing that I should have changed `count` to `length` here!

**Pull request merged**